File: android/app/src/main/java/dev/anuz/quickinbox/ui/compose/ComposeViewModel.kt
package dev.anuz.quickinbox.ui.compose

import android.content.Context
import android.net.Uri
import android.provider.OpenableColumns
import android.util.Base64
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.viewModelScope
import dev.anuz.quickinbox.data.AppPreferences
import dev.anuz.quickinbox.data.QuickInboxApi
import dev.anuz.quickinbox.domain.ComposeMessage
import dev.anuz.quickinbox.domain.MailAddress
import dev.anuz.quickinbox.domain.ForwardMessage
import dev.anuz.quickinbox.domain.OutboundAttachment
import dev.anuz.quickinbox.domain.ReplyMessage
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.util.UUID

sealed class ComposeMode {
    data object NewMessage : ComposeMode()
    data class Draft(val id: String) : ComposeMode()
    data class Reply(
        val messageId: String,
        val recipient: String,
        val subject: String,
        val recipientName: String? = null,
        val fromAddressHint: String? = null,
        val originalTo: String = "",
        val originalCc: String? = null
    ) : ComposeMode()
    sealed class Forward : ComposeMode() {
        abstract val subject: String
        abstract val messageCount: Int
        abstract val attachmentCount: Int
        abstract val fromAddressHint: String?

        data class Message(
            val messageId: String,
            override val subject: String,
            override val attachmentCount: Int = 0,
            override val fromAddressHint: String? = null
        ) : Forward() {
            override val messageCount: Int = 1
        }

        data class Thread(
            val threadId: String,
            override val subject: String,
            override val messageCount: Int,
            override val attachmentCount: Int = 0,
            override val fromAddressHint: String? = null
        ) : Forward()
    }

    val title: String
        get() = when (this) {
            NewMessage -> "New message"
            is Draft -> "Draft"
            is Reply -> "Reply"
            is Forward.Message -> "Forward"
            is Forward.Thread -> "Forward all"
        }

    val draftId: String? get() = (this as? Draft)?.id
}

data class ComposeAttachment(
    val id: String = UUID.randomUUID().toString(),
    val filename: String,
    val contentType: String,
    val byteCount: Int,
    val encodedContent: String
) {
    val outbound get() = OutboundAttachment(filename, contentType, encodedContent)
}

data class ComposeUiState(
    val to: List<String> = emptyList(),
    val cc: List<String> = emptyList(),
    val bcc: List<String> = emptyList(),
    val subject: String = "",
    val body: String = "",
    val addresses: List<MailAddress> = emptyList(),
    val selectedFromAddressId: String? = null,
    val attachments: List<ComposeAttachment> = emptyList(),
    val includeOriginalAttachments: Boolean = true,
    val isLoadingAddresses: Boolean = false,
    val isLoadingDraft: Boolean = false,
    val isImporting: Boolean = false,
    val isSending: Boolean = false,
    val errorMessage: String? = null,
    val attachmentMessage: String? = null
)

class ComposeViewModel(
    private val api: QuickInboxApi,
    val mode: ComposeMode,
    private val preferences: AppPreferences
) : ViewModel() {
    private val _state = MutableStateFlow(
        when (mode) {
            ComposeMode.NewMessage -> ComposeUiState()
            is ComposeMode.Draft -> ComposeUiState()
            is ComposeMode.Reply -> ComposeUiState(to = listOf(mode.recipient), subject = mode.subject)
            is ComposeMode.Forward -> ComposeUiState(subject = mode.subject)
        }
    )
    val state = _state.asStateFlow()

    private var didLoadAddresses = false
    private var didLoadDraft = false

    val isReply: Boolean get() = mode is ComposeMode.Reply

    fun onTo(value: List<String>) = _state.update { it.copy(to = value, errorMessage = null) }
    fun onCc(value: List<String>) = _state.update { it.copy(cc = value, errorMessage = null) }
    fun onBcc(value: List<String>) = _state.update { it.copy(bcc = value, errorMessage = null) }
    fun onSubject(value: String) = _state.update { it.copy(subject = value, errorMessage = null) }
    fun onBody(value: String) = _state.update { it.copy(body = value, errorMessage = null) }
    fun onFrom(id: String) = _state.update { it.copy(selectedFromAddressId = id) }
    fun onIncludeOriginalAttachments(include: Boolean) =
        _state.update { it.copy(includeOriginalAttachments = include, errorMessage = null) }

    fun load() {
        viewModelScope.launch {
            loadAddressesIfNeeded()
            loadDraftIfNeeded()
        }
    }

    fun importUris(context: Context, uris: List<Uri>) {
        if (uris.isEmpty()) return
        viewModelScope.launch {
            _state.update { it.copy(isImporting = true, attachmentMessage = null) }
            val imported = _state.value.attachments.toMutableList()
            val messages = mutableListOf<String>()
            try {
                for (uri in uris) {
                    if (imported.size >= MAX_COUNT) {
                        messages += "You can attach up to $MAX_COUNT files."
                        break
                    }
                    try {
                        val attachment = withContext(Dispatchers.IO) { readAttachment(context, uri) }
                        if (imported.sumOf { it.byteCount } + attachment.byteCount > MAX_TOTAL_BYTES) {
                            messages += "Attachments cannot exceed 25 MB in total."
                            continue
                        }
                        imported += attachment
                    } catch (error: CancellationException) {
                        throw error
                    } catch (error: Exception) {
                        messages += error.message ?: "A file could not be attached."
                    }
                }
                _state.update {
                    it.copy(
                        attachments = imported,
                        attachmentMessage = messages.takeIf { list -> list.isNotEmpty() }?.joinToString("\n")
                    )
                }
            } finally {
                _state.update { it.copy(isImporting = false) }
            }
        }
    }

    fun removeAttachment(id: String) {
        if (_state.value.isSending) return
        _state.update { it.copy(attachments = it.attachments.filterNot { attachment -> attachment.id == id }, attachmentMessage = null) }
    }

    fun send(onSent: () -> Unit) {
        val snapshot = _state.value
        if (snapshot.isSending) return
        val validation = validationMessage(snapshot)
        if (validation != null) {
            _state.update { it.copy(errorMessage = validation) }
            return
        }
        viewModelScope.launch {
            _state.update { it.copy(isSending = true, errorMessage = null) }
            try {
                val outbound = snapshot.attachments.takeIf { it.isNotEmpty() }?.map { it.outbound }
                withContext(Dispatchers.IO) {
                    when (val current = mode) {
                        ComposeMode.NewMessage, is ComposeMode.Draft -> api.send(
                            ComposeMessage(
                                draftId = current.draftId,
                                fromAddressId = snapshot.selectedFromAddressId,
                                to = snapshot.to.joinToString(", "),
                                cc = snapshot.cc.joinToString(", ").ifEmpty { null },
                                bcc = snapshot.bcc.joinToString(", ").ifEmpty { null },
                                subject = snapshot.subject.trim(),
                                text = snapshot.body.trim(),
                                attachments = outbound
                            )
                        )
                        is ComposeMode.Reply -> api.reply(
                            current.messageId,
                            ReplyMessage(
                                fromAddressId = snapshot.selectedFromAddressId,
                                to = snapshot.to.joinToString(", ").ifEmpty { null },
                                cc = snapshot.cc.joinToString(", ").ifEmpty { null },
                                bcc = snapshot.bcc.joinToString(", ").ifEmpty { null },
                                text = snapshot.body.trim(),
                                attachments = outbound
                            )
                        )
                        is ComposeMode.Forward -> {
                            val request = ForwardMessage(
                                fromAddressId = snapshot.selectedFromAddressId,
                                to = snapshot.to.joinToString(", "),
                                cc = snapshot.cc.joinToString(", ").ifEmpty { null },
                                bcc = snapshot.bcc.joinToString(", ").ifEmpty { null },
                                text = snapshot.body.trim().ifEmpty { null },
                                includeAttachments = snapshot.includeOriginalAttachments
                            )
                            when (current) {
                                is ComposeMode.Forward.Message -> api.forwardMessage(current.messageId, request)
                                is ComposeMode.Forward.Thread -> api.forwardThread(current.threadId, request)
                            }
                        }
                    }
                }
                onSent()
            } catch (error: CancellationException) {
                throw error
            } catch (error: Exception) {
                _state.update { it.copy(errorMessage = error.message ?: "QuickInbox could not send this message. Try again.") }
            } finally {
                _state.update { it.copy(isSending = false) }
            }
        }
    }

    private suspend fun loadAddressesIfNeeded() {
        if (mode is ComposeMode.Forward) return
        if (didLoadAddresses) return
        didLoadAddresses = true
        _state.update { it.copy(isLoadingAddresses = true) }
        try {
            val loaded = withContext(Dispatchers.IO) { api.addresses() }
            _state.update { state ->
                val selected = if (mode is ComposeMode.Reply || mode is ComposeMode.Forward) {
                    state.selectedFromAddressId
                } else {
                    state.selectedFromAddressId ?: preferredAddress(loaded)?.id
                }
                state.copy(
                    addresses = loaded,
                    selectedFromAddressId = selected,
                    isLoadingAddresses = false,
                    errorMessage = if (loaded.isEmpty()) {
                        "No sending address is configured. Add one in QuickInbox settings first."
                    } else state.errorMessage
                )
            }
        } catch (error: CancellationException) {
            throw error
        } catch (error: Exception) {
            _state.update { it.copy(isLoadingAddresses = false, errorMessage = error.message ?: "Sending addresses could not be loaded.") }
        } finally {
            _state.update { it.copy(isLoadingAddresses = false) }
        }
    }

    private suspend fun loadDraftIfNeeded() {
        val draftId = mode.draftId ?: return
        if (didLoadDraft) return
        didLoadDraft = true
        _state.update { it.copy(isLoadingDraft = true) }
        try {
            val draft = withContext(Dispatchers.IO) { api.draft(draftId) }
            _state.update { state ->
                state.copy(
                    to = parseRecipients(draft.to),
                    cc = parseRecipients(draft.cc.orEmpty()),
                    bcc = parseRecipients(draft.bcc.orEmpty()),
                    subject = draft.subject,
                    body = draft.text.orEmpty(),
                    selectedFromAddressId = state.addresses.firstOrNull {
                        it.address.equals(draft.fromAddress, ignoreCase = true)
                    }?.id ?: preferredAddress(state.addresses)?.id,
                    isLoadingDraft = false
                )
            }
        } catch (error: CancellationException) {
            throw error
        } catch (error: Exception) {
            _state.update {
                it.copy(
                    isLoadingDraft = false,
                    errorMessage = error.message ?: "QuickInbox couldn’t load this draft. Try again."
                )
            }
        } finally {
            _state.update { it.copy(isLoadingDraft = false) }
        }
    }

    private fun preferredAddress(addresses: List<MailAddress>): MailAddress? {
        val saved = preferences.selectedSendingAddressId
        return addresses.firstOrNull { it.id == saved }
            ?: addresses.firstOrNull { it.isDefault }
            ?: addresses.firstOrNull()
    }

    private fun validationMessage(state: ComposeUiState): String? {
        if (mode !is ComposeMode.Forward && state.body.trim().isEmpty()) return "Write a message before sending."
        if (mode !is ComposeMode.Reply && mode !is ComposeMode.Forward && state.selectedFromAddressId == null) {
            return "Choose a sending address."
        }
        recipientError(state.to, "To", required = true)?.let { return it }
        recipientError(state.cc, "Cc", required = false)?.let { return it }
        recipientError(state.bcc, "Bcc", required = false)?.let { return it }
        if (mode is ComposeMode.Reply) return null
        val subject = state.subject.trim()
        if (subject.isEmpty()) return "Add a subject."
        if (subject.length > 200) return "The subject must be 200 characters or fewer."
        return null
    }

    private fun parseRecipients(value: String): List<String> =
        value.split(',').map { it.trim() }.filter { it.isNotEmpty() }

    private fun recipientError(recipients: List<String>, label: String, required: Boolean): String? {
        if (recipients.isEmpty()) {
            return if (required) "Add at least one recipient." else null
        }
        for (recipient in recipients) {
            if (!isValidEmail(recipient)) return "Check the email address in $label: $recipient"
        }
        return null
    }

    private fun isValidEmail(value: String): Boolean {
        if (value.any { it.isWhitespace() }) return false
        val parts = value.split('@')
        if (parts.size != 2 || parts[0].isEmpty() || parts[1].isEmpty()) return false
        return parts[1].contains('.') && !parts[1].startsWith('.') && !parts[1].endsWith('.')
    }

    private fun readAttachment(context: Context, uri: Uri): ComposeAttachment {
        val resolver = context.contentResolver
        var filename = "attachment"
        var size = 0L
        resolver.query(uri, null, null, null, null)?.use { cursor ->
            val nameIndex = cursor.getColumnIndex(OpenableColumns.DISPLAY_NAME)
            val sizeIndex = cursor.getColumnIndex(OpenableColumns.SIZE)
            if (cursor.moveToFirst()) {
                if (nameIndex >= 0) filename = cursor.getString(nameIndex) ?: filename
                if (sizeIndex >= 0) size = cursor.getLong(sizeIndex)
            }
        }
        if (size > MAX_BYTES) throw IllegalArgumentException("$filename exceeds the 5 MB limit.")
        val bytes = resolver.openInputStream(uri)?.use { it.readBytes() }
            ?: throw IllegalArgumentException("$filename could not be attached.")
        if (bytes.isEmpty()) throw IllegalArgumentException("$filename is empty.")
        if (bytes.size > MAX_BYTES) throw IllegalArgumentException("$filename exceeds the 5 MB limit.")
        val type = resolver.getType(uri) ?: "application/octet-stream"
        return ComposeAttachment(
            filename = filename,
            contentType = type,
            byteCount = bytes.size,
            encodedContent = Base64.encodeToString(bytes, Base64.NO_WRAP)
        )
    }

    companion object {
        const val MAX_COUNT = 5
        const val MAX_BYTES = 5 * 1024 * 1024
        const val MAX_TOTAL_BYTES = 25 * 1024 * 1024
    }
}

class ComposeViewModelFactory(
    private val api: QuickInboxApi,
    private val mode: ComposeMode,
    private val preferences: AppPreferences
) : ViewModelProvider.Factory {
    @Suppress("UNCHECKED_CAST")
    override fun <T : ViewModel> create(modelClass: Class<T>): T =
        ComposeViewModel(api, mode, preferences) as T
}

File: ios/quickinbox/quickinbox/Features/Compose/ComposeViewModel.swift
import Combine
import Foundation
import UniformTypeIdentifiers

nonisolated enum ComposeMode: Equatable, Sendable {
    case newMessage
    case draft(draftID: String)
    case reply(messageID: String, recipient: String, subject: String)
    case forward(subject: String, body: String)

    var navigationTitle: String {
        switch self {
        case .newMessage: "New Message"
        case .draft: "Draft"
        case .reply: "Reply"
        case .forward: "Forward"
        }
    }

    var draftID: String? {
        if case .draft(let draftID) = self { draftID } else { nil }
    }
}

nonisolated struct ComposeAttachment: Identifiable, Equatable, Sendable {
    let id: UUID
    let filename: String
    let contentType: String
    let byteCount: Int
    let encodedContent: String

    init(
        id: UUID = UUID(),
        filename: String,
        contentType: String,
        byteCount: Int,
        encodedContent: String
    ) {
        self.id = id
        self.filename = filename
        self.contentType = contentType
        self.byteCount = byteCount
        self.encodedContent = encodedContent
    }

    var outboundValue: OutboundAttachment {
        OutboundAttachment(filename: filename, type: contentType, content: encodedContent)
    }
}

@MainActor
final class ComposeViewModel: ObservableObject {
    nonisolated static let maximumAttachmentCount = 5
    nonisolated static let maximumAttachmentBytes = 5 * 1_024 * 1_024
    nonisolated static let maximumTotalAttachmentBytes = 25 * 1_024 * 1_024

    @Published var to = ""
    @Published var cc = ""
    @Published var bcc = ""
    @Published var subject = ""
    @Published var body = ""
    @Published private(set) var addresses: [MailAddress]
    @Published var selectedFromAddressID: String?
    @Published private(set) var attachments: [ComposeAttachment] = []
    @Published private(set) var isLoadingAddresses = false
    @Published private(set) var isLoadingDraft = false
    @Published private(set) var isImportingAttachments = false
    @Published private(set) var isSending = false
    @Published var errorMessage: String?
    @Published var attachmentMessage: String?

    let mode: ComposeMode

    private let api: QuickInboxAPI
    private var didAttemptAddressLoad = false
    private var didAttemptDraftLoad = false

    init(api: QuickInboxAPI, mode: ComposeMode, addresses: [MailAddress] = []) {
        self.api = api
        self.mode = mode
        self.addresses = addresses

        switch mode {
        case .newMessage:
            selectedFromAddressID = Self.preferredAddress(in: addresses)?.id
        case .draft:
            selectedFromAddressID = nil
        case .reply(_, let recipient, let subject):
            to = recipient
            self.subject = subject
            // A nil selection lets the server reply from the mailbox that received
            // the original, including a catch-all address that is not saved.
            selectedFromAddressID = nil
        case .forward(let subject, let body):
            self.subject = subject
            self.body = body
            selectedFromAddressID = Self.preferredAddress(in: addresses)?.id
        }
    }

    var isReply: Bool {
        if case .reply = mode { true } else { false }
    }

    var isDraft: Bool { mode.draftID != nil }

    var totalAttachmentBytes: Int {
        attachments.reduce(0) { $0 + $1.byteCount }
    }

    var hasUnsavedChanges: Bool {
        !body.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
            || !attachments.isEmpty
            || (!isReply && (
                !to.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
                    || !cc.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
                    || !bcc.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
                    || !subject.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
            ))
    }

    var canSend: Bool {
        !isSending && !isImportingAttachments && !isLoadingDraft && validationMessage == nil
    }

    var validationMessage: String? {
        let trimmedBody = body.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmedBody.isEmpty else { return "Write a message before sending." }

        switch mode {
        case .newMessage, .draft, .forward:
            guard selectedFromAddressID != nil else {
                return "Choose a sending address."
            }
            if let error = Self.recipientError(to, label: "To", required: true) { return error }
            if let error = Self.recipientError(cc, label: "Cc", required: false) { return error }
            if let error = Self.recipientError(bcc, label: "Bcc", required: false) { return error }

            let trimmedSubject = subject.trimmingCharacters(in: .whitespacesAndNewlines)
            guard !trimmedSubject.isEmpty else { return "Add a subject." }
            guard trimmedSubject.count <= 200 else {
                return "The subject must be 200 characters or fewer."
            }
        case .reply:
            break
        }

        return nil
    }

    func loadAddressesIfNeeded() async {
        guard !didAttemptAddressLoad else { return }
        didAttemptAddressLoad = true
        await loadAddresses()
    }

    func loadDraftIfNeeded() async {
        guard !didAttemptDraftLoad, let draftID = mode.draftID else { return }
        didAttemptDraftLoad = true
        isLoadingDraft = true
        defer { isLoadingDraft = false }

        do {
            let draft = try await api.draft(id: draftID)

            to = draft.to
            cc = draft.cc ?? ""
            bcc = draft.bcc ?? ""
            subject = draft.subject
            body = draft.text ?? ""
            selectedFromAddressID = addresses.first(where: {
                $0.address.caseInsensitiveCompare(draft.fromAddress) == .orderedSame
            })?.id ?? Self.preferredAddress(in: addresses)?.id
        } catch {
            guard !Self.isUnauthorized(error) else { return }
            errorMessage = (error as? LocalizedError)?.errorDescription
                ?? "QuickInbox couldn’t load this draft. Try again."
        }
    }

    func retryLoadingAddresses() async {
        await loadAddresses()
    }

    func importFiles(_ result: Result<[URL], Error>) async {
        attachmentMessage = nil

        let urls: [URL]
        do {
            urls = try result.get()
        } catch {
            // Cancellation is not an error that needs to replace the draft state.
            if (error as NSError).code != NSUserCancelledError {
                attachmentMessage = "The selected files could not be opened."
            }
            return
        }

        guard !urls.isEmpty else { return }
        isImportingAttachments = true
        defer { isImportingAttachments = false }

        var imported = attachments
        var messages: [String] = []

        for url in urls {
            guard imported.count < Self.maximumAttachmentCount else {
                messages.append("You can attach up to \(Self.maximumAttachmentCount) files.")
                break
            }

            do {
                let attachment = try await Self.readAttachment(from: url)
                guard imported.reduce(0, { $0 + $1.byteCount }) + attachment.byteCount
                        <= Self.maximumTotalAttachmentBytes else {
                    messages.append("Attachments cannot exceed 25 MB in total.")
                    continue
                }
                imported.append(attachment)
            } catch let error as AttachmentImportError {
                messages.append(error.localizedDescription)
            } catch {
                messages.append("\(url.lastPathComponent) could not be attached.")
            }
        }

        attachments = imported
        attachmentMessage = messages.isEmpty ? nil : messages.joined(separator: "\n")
    }

    func removeAttachment(id: ComposeAttachment.ID) {
        guard !isSending else { return }
        attachments.removeAll { $0.id == id }
        attachmentMessage = nil
    }

    func send() async -> SendMessageResponse? {
        // This guard is the final duplicate-submit barrier, independent of the UI.
        guard !isSending else { return nil }
        guard let validationMessage else {
            errorMessage = nil
            isSending = true
            defer { isSending = false }

            do {
                let response: SendMessageResponse
                let outboundAttachments = attachments.isEmpty
                    ? nil
                    : attachments.map(\.outboundValue)

                switch mode {
                case .newMessage, .draft, .forward:
                    response = try await api.send(
                        ComposeMessage(
                            draftID: mode.draftID,
                            fromAddressID: selectedFromAddressID,
                            to: to.trimmingCharacters(in: .whitespacesAndNewlines),
                            cc: Self.nilIfEmpty(cc),
                            bcc: Self.nilIfEmpty(bcc),
                            subject: subject.trimmingCharacters(in: .whitespacesAndNewlines),
                            text: body.trimmingCharacters(in: .whitespacesAndNewlines),
                            html: nil,
                            attachments: outboundAttachments
                        )
                    )
                case .reply(let messageID, _, _):
                    response = try await api.reply(
                        to: messageID,
                        message: ReplyMessage(
                            fromAddressID: selectedFromAddressID,
                            text: body.trimmingCharacters(in: .whitespacesAndNewlines),
                            html: nil,
                            attachments: outboundAttachments
                        )
                    )
                }

                AppFeedback.play(.messageSent)
                return response
            } catch {
                guard !Self.isUnauthorized(error) else { return nil }
                // Keep all entered fields and attachments intact so Retry is safe.
                errorMessage = (error as? LocalizedError)?.errorDescription
                    ?? "QuickInbox could not send this message. Try again."
                AppFeedback.error()
                return nil
            }
        }

        errorMessage = validationMessage
        AppFeedback.play(.warning)
        return nil
    }

    private func loadAddresses() async {
        guard !isLoadingAddresses else { return }
        isLoadingAddresses = true
        defer { isLoadingAddresses = false }

        do {
            let loaded = try await api.addresses()
            addresses = loaded
            if !isReply, selectedFromAddressID == nil {
                selectedFromAddressID = Self.preferredAddress(in: loaded)?.id
            }
            errorMessage = loaded.isEmpty
                ? "No sending address is configured. Add one in QuickInbox settings first."
                : nil
        } catch {
            guard !Self.isUnauthorized(error) else { return }
            errorMessage = (error as? LocalizedError)?.errorDescription
                ?? "Sending addresses could not be loaded."
        }
    }

    private static func isUnauthorized(_ error: Error) -> Bool {
        if let apiError = error as? APIError, apiError == .unauthorized { return true }
        return false
    }

    private static func preferredAddress(in addresses: [MailAddress]) -> MailAddress? {
        if let savedID = UserDefaults.standard.string(forKey: AppPreferences.selectedSendingAddressID),
           let saved = addresses.first(where: { $0.id == savedID }) {
            return saved
        }
        return addresses.first(where: \.isDefault) ?? addresses.first
    }

    private static func nilIfEmpty(_ value: String) -> String? {
        let trimmed = value.trimmingCharacters(in: .whitespacesAndNewlines)
        return trimmed.isEmpty ? nil : trimmed
    }

    private static func recipientError(
        _ value: String,
        label: String,
        required: Bool
    ) -> String? {
        let recipients = value
            .split(separator: ",", omittingEmptySubsequences: false)
            .map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }

        if recipients.count == 1, recipients[0].isEmpty {
            return required ? "Add at least one recipient." : nil
        }
        if recipients.contains(where: \.isEmpty) {
            return "Remove the empty address from \(label)."
        }

        for recipient in recipients where !isValidEmail(recipient) {
            return "Check the email address in \(label): \(recipient)"
        }
        return nil
    }

    private static func isValidEmail(_ value: String) -> Bool {
        guard !value.contains(where: \.isWhitespace) else { return false }
        let parts = value.split(separator: "@", omittingEmptySubsequences: false)
        guard parts.count == 2, !parts[0].isEmpty, !parts[1].isEmpty else { return false }
        return parts[1].contains(".") && !parts[1].hasPrefix(".") && !parts[1].hasSuffix(".")
    }

    private nonisolated static func readAttachment(from url: URL) async throws -> ComposeAttachment {
        try await Task.detached(priority: .userInitiated) {
            let accessed = url.startAccessingSecurityScopedResource()
            defer {
                if accessed { url.stopAccessingSecurityScopedResource() }
            }

            let values = try url.resourceValues(forKeys: [
                .contentTypeKey,
                .fileSizeKey,
                .isRegularFileKey,
                .nameKey
            ])
            let filename = values.name ?? url.lastPathComponent
            guard values.isRegularFile != false else {
                throw AttachmentImportError.notAFile(filename)
            }
            if let size = values.fileSize, size > maximumAttachmentBytes {
                throw AttachmentImportError.tooLarge(filename)
            }

            let data = try Data(contentsOf: url, options: [.mappedIfSafe])
            guard !data.isEmpty else { throw AttachmentImportError.empty(filename) }
            guard data.count <= maximumAttachmentBytes else {
                throw AttachmentImportError.tooLarge(filename)
            }

            return ComposeAttachment(
                filename: filename.isEmpty ? "attachment" : filename,
                contentType: values.contentType?.preferredMIMEType ?? "application/octet-stream",
                byteCount: data.count,
                encodedContent: data.base64EncodedString()
            )
        }.value
    }
}

private nonisolated enum AttachmentImportError: LocalizedError {
    case empty(String)
    case notAFile(String)
    case tooLarge(String)

    var errorDescription: String? {
        switch self {
        case .empty(let filename): "\(filename) is empty."
        case .notAFile(let filename): "\(filename) is not a file."
        case .tooLarge(let filename): "\(filename) exceeds the 5 MB limit."
        }
    }
}
